### ISSUE TITLE
Include pregenerated configure file on in Github tarball

### ISSUE THREAD

**Issue by noocsharp** [htop-dev/htop#618]
The release tarball on Bintray used to include a pregenerated configure file. However now that that seems to be gone we have to use the Github tarball, which doesn't.

This lack of a configure file also causes the instructions for building here: https://htop.dev/downloads.html#sources to be incorrect.

**Comment by fasterit** [comment]
@natoscott commented in https://github.com/htop-dev/htop/issues/620#issuecomment-832396011 that we'll not be providing configure scripts inside tarballs anymore. So this issue should probably be sorted by fixing the build instructions to match.

**Comment by noocsharp** [comment]
Is there any way for not providing configure scripts to be reconsidered? My understanding that release tarballs are supposed to include configure scripts to allow the program to be installed without worrying about autotools. It also seems trivial to generate a release tarball by just using `make dist`.

**Comment by fasterit** [comment]
Do you have any rationale for having make, gcc and friends but not autotools on a platform?

We discussed that in the core team but found no real reasoning any more in 2021. Of course it was different when people were mailing around floppys or the lucky ones downloading from Simtel, but those days are gone. Typically it is just one `apt get autoconf` away.

We'd be happy to reconsider the decision if there are good arguments to do so.

**Comment by noocsharp** [comment]
make, a c compiler (tcc also builds htop just fine) and friends are the lowest common denominator. They can be found on any standard system, and don't have any dependencies. If you force the user to have autotools installed, you also force them to have perl installed. So whether intentional or not, you now need perl to build htop.

I think the question should be why _should_ autotools be required for the user to build a program when it is so easy to avoid it? I'm not saying y'all should ditch autotools as your build system, all I ask for is a tarball with a prebuilt configure file. Autotools is designed to be used like this, and nearly all other projects that use autotools as their build system ship their tarballs with a configure file.

**Comment by BenBE** [comment]
I've built `htop` on systems like YoctoLinux and various other platforms where you basically first have to build your own toolchain before even getting to building things like `htop`. And by the time you could build `htop` on those platforms you already had a bunch of other tools built that essentially suffice to also build `autotools`. So even while having `autotools` as a dependency for builds might seem like adding in much trouble, it's not actually much of a new dependency: You can't essentially have e.g. a Debian system without `perl`. And even with things like OpenWRT there's several places where at least for the host (build) environment you'll at some point invoke some perl script or another to build some stuff.

**Comment by noocsharp** [comment]
Suppose I have built a system where I've gone through the effort to avoid things like autotools and perl, even if it is purely for aesthetic reasons. If a `configure` file is included, I can compile `htop` on my system. If it isn't, then I have to write my own Makefile or something. It's as simple as that.

As I said above, autotools are designed in such a way they don't need to be installed in order to build a program. By requiring the user to install autotools, you are undermining a major advantage of autotools compared to something like meson, where python becomes a build dependency for a program. In my opinion, its a little bit absurd to have perl as a build dependency for a program when it can be avoided by adding a single file to the release tarball, especially since the build system is explicitly designed to be used like this.

If this line of reasoning doesn't convince you, I'm not sure that anything else that I say will. So feel free to close the issue if you are still unconvinced. Thanks for taking the time to discuss this.

**Comment by cassioneri** [comment]
I just want to second everything @noocsharp said and also ask you to reconsider your decision.

**Comment by fasterit** [comment]
We are still looking for compelling reasons to have the autotools output in the repo. Does anybody have an argument beyond historic preferences?

**Comment by noocsharp** [comment]
I don't think anybody is advocating for autotools output in the repo. Just the release tarball. I agree that it's not a good idea to put autotools output in the git repo.

**Comment by Explorer09** [comment]
How about thinking the other way: Removing autotools-generated scripts from release tarball brings no advantage other than adding a dependency hell to people who want to build htop. Why bother removing in the first place? (A release tarball is not a version controlled repo - you don't need to regenerate the scripts periodically to keep them up to date; you only regenerate them once per release.)

**Comment by BenBE** [comment]
The problem is a bit the other way around really: On platforms where we can provide the configure files you usually also have a package manager providing you easy access to install autoconf. On the other hand, for situations where the pre-generated configure doesn't work (I think of cross-compiling in certain situations) you need to regenerate the configure for your needs anyways - and in that case you probably have enough of a compile environment in place anyway so that having autoconf doesn't add much overhead anymore.

**Comment by Explorer09** [comment]
On Tue, May 25, 2021 at 1:29 PM BenBE ***@***.***> wrote:
>
> On platforms where we can provide the configure files you usually also have a package manager providing you easy access to install autoconf.

The keyword is "usually". It's not necessarily true for small,
embedded systems, where there's no package manager and the toolchain
has to be almost built manually.

> On the other hand, for situations where the pre-generated configure doesn't work (I think of cross-compiling in certain situations) you need to regenerate the configure for your needs anyways - and in that case you probably have enough of a compile environment in place anyway so that having autoconf doesn't add much overhead anymore.

What I'm expecting is that configure won't need to be regenerated
unless there are changes in the build system, or addition or removal
of source files. The cases where the pre-generated configure doesn't
work are only a minority. It's a bad idea to sacrifice the majority of
cases where it does work.

And by the way, there's another compelling reason to include the
configure: Many autotools/M4 macros many project made are not
future-proof - autotools are known for introducing breaking changes
throughout major versions (check their NEWS files to see what I mean)
and macros written for one version might not be compatible with the
later version. Project like GCC (as far as I know) requires a specific
version of autoconf (not any newer or older version) to generate its
scripts for this reason.

**Comment by BenBE** [comment]
> > On Tue, May 25, 2021 at 1:29 PM BenBE ***@***.***> wrote: On platforms where we can provide the configure files you usually also have a package manager providing you easy access to install autoconf.
>
> The keyword is "usually". It's not necessarily true for small, embedded systems, where there's no package manager and the toolchain has to be almost built manually.

Yes, and that's where I say from my experience, that once you have such an embedded system build system, you usually also have to build autoconf for it anyway.

**Comment by badger200** [comment]
@fasterit @BenBE I am attempting to compile htop on iOS/Darwin (jailbroken) as the current binary is 2.20. 

My problem is "missing Makefile.in" and "missing config.h.in". I have never had these issues before and I've successfully compiled at least a dozen projects using GNU autotools and the ./configure scripts. 

I have never seen a project use ./configure but not containing a Makefile.in and config.h.in. You guys are definitely doing something highly abnormal, but I am not experienced enough to identify it more precisely than this. 

Looking forward to compiling this project big time!!! Love that you added disk I/O and network traffic!! A dream come true (hopefully to come true)

**Comment by BenBE** [comment]
@badger200 Just run [autogen.sh](https://github.com/htop-dev/htop/blob/master/autogen.sh) (or the one single command contained in it) to generate all those files. This will generate both `configure` from `configure.ac` and `Makefile.in` from `Makefile.am`.

**Comment by badger200** [comment]
@BenBE I’ve isolated the issue and it’s due to the Darwin/iOS factory security enforcements, breaking “shebangs” (#!/bin/sh) triggering dmesg “Sandbox: Deny perl(1234): process-exec-interpreter”. 

There are some clunky workarounds but what this boils down to is the guy a few comments above me was right when he said requiring Perl on all systems instead of aiding those better suited to standard ./configure behavior is more trouble than its worth. 

In other words, I’ve never seen any other project use this arbitrary “autogen.sh” script you have, it seems like you’re doing something not “kosher” with the GNU automake system and that’s the root cause of these issues on various systems.
